Allow samplers to be initialized without target

nabriis opened the pull request:
Closes #398 (Target=None)
Closes #397 (Proposal based refactor)
Closes #414 (Reset method)
Closes #401 (Initial point in sample)

**Change overview**
1) Added _initialize abstractmethod that all subclasses must implement.
This method is called before sampling and allows "lazy" initialization allowing target not be set during init.
2) Updated all samplers to use _initialize and allow Target=None.
3) Refactoring proposal implementation to be aligned with target (including validate_proposal)
4) Fixed the reset method to properly reset using the new _initialize method
5) Removed initial point from samples
6) Added gitignore on CUQI_samples

**Area of interest for reviewer**
- For the NUTS sampler, I was unable to completely avoid the _pre_sample and _pre_warmup methods because NUTS acts different whether you start sampling first or start warm_up first. We should look into this. If fixed, I believe we can get rid of _pre_sample and _pre_warmup and use _initialize instead.
- The _scale_temp variables in MH and CWMH is starting to annoy me. I think we should fix the bug in sampler main and get rid of these variables in our new samplers.

amal-ghamdi commented on tests/zexperimental/test_mcmc.py:
```diff
@@ -478,3 +478,18 @@ def test_state_is_fully_updated_after_warmup_step(sampler: cuqi.experimental.mcm
+]
+
+@pytest.mark.parametrize("sampler_class", target_None_Samplers)
+def test_target_None_init_in_samplers(sampler_class):
```
Nice test :)

amal-ghamdi reviewed (changes requested):
Thank you so much @nabriis for this refactor/feature. It is nice to see much improved code for setting targets and proposals! And increased consistency overall (initial_point, etc). I just made a few minor comments.

nabriis commented on .gitignore:
```diff
@@ -32,3 +32,6 @@ dist/
 *.pickle
+
+# Ignore certain files from demos
+/demos/CUQI_samples/*
```
Its because its the default name for batching. Perhaps */CUQI_samples* ?

nabriis commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -67,72 +67,65 @@ class CWMHNew(ProposalBasedSamplerNew):
     def scale(self):
         """ Get the scale parameter. """
+        if isinstance(self._scale, Number):
+            self._scale = np.ones(self.dim)*self._scale
```
Because target might not be set so dim not available

amal-ghamdi replied:
got it!

nabriis commented on tests/zexperimental/test_mcmc.py:
```diff
@@ -319,7 +319,7 @@ def test_NUTS_regression_warmup(target: cuqi.density.Density):
     cuqi.experimental.mcmc.CWMHNew,
     cuqi.experimental.mcmc.RegularizedLinearRTONew, # Due to the _choose_stepsize method
-    cuqi.experimental.mcmc.NUTSNew
+    cuqi.experimental.mcmc.NUTSNew,
```
Nothing. It was after stripping the Gibbs testing out (which was a new line)

amal-ghamdi replied:
thanks :)

nabriis commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -67,72 +67,65 @@ class CWMHNew(ProposalBasedSamplerNew):
+        self.scale_temp = None
 
-        # set initial acceptance rate
+    def _pre_warmup(self):
+        super()._pre_warmup()
         self._acc = [np.ones((self.dim))]
```
Good point. Think we need self._initialize method that is called only once, but after target is set. I will look into it

nabriis commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -67,72 +67,65 @@ class CWMHNew(ProposalBasedSamplerNew):
-            self._scale = np.ones(self.dim)*value
-        elif isinstance(value, np.ndarray):
-            self._scale = value
-        self._scale_temp = self._scale.copy()
```
Ah yes. Fixed. We really need to get rid of this scale temp though..

amal-ghamdi replied:
Agree! we need to look into if scale temp is needed.

amal-ghamdi replied:
Just one comment here, when CWMH sets the scale in `tune`,  the value of setting the scale depends on `scale_temp`.  
```
        # Update the scale parameter
        self.scale = np.minimum(scale_temp, np.ones(self.dim))
```

so for example if 
```
sampler.scale #is 0.9, and
sampler._scale_temp is 0.9
```
then if the user changed `sampler.scale` to be 0.1
then `sampler._scale_temp` which is used to determine the value in the line from `tune` above will be 0.9

nabriis replied:
In CWMH a local variable for scale_temp is used ss shown in your example. The previous self._scale_temp is used to define that variable early in tune. As the last step in tune self._scale_temp is updated again.

I think this is fine and in line with the other samplers.

Anyway I suggest we look into getting rid of this scale temp in general.

amal-ghamdi replied:
It is probably just fine to keep it like it is, just wanted to mention this scenario where there might be some inconsistency between the `scale` and `scale_temp`
```

sampler = CWMH(target, scale=0.9) 
sampler.warmup(3)
sampler.scale = 0.1
sampler.tune()
# although tune is probably not meant to be called by user, which is another issue 
# if we should make it hidden: `_tune`
# now scale_temp that is used to set the local scale_temp in tune is derived from scale=0.9
```

nabriis replied:
Agreed. Anyway my suggestion is to get rid of scale_temp if possible.

nabriis commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -67,72 +67,65 @@ class CWMHNew(ProposalBasedSamplerNew):
                 mean=lambda location: location,
                 std=lambda scale: scale,
                 geometry=self.dim,
             )
```
Yes :) Good catch

nabriis commented on cuqi/experimental/mcmc/_hmc.py:
```diff
@@ -289,32 +305,14 @@ def _pre_warmup(self):
-            self._mu = np.log(10*self._epsilon)
-
-        if self._epsilon_bar is None: # Initial value of epsilon_bar
+        if self._epsilon_bar == "unset": # Initial value of epsilon_bar for tuning
```
I could not get rid of pre_warmup and pre_sample in NUTS entirely because there is different behavior for _epsilon_bar depending if you do sample or warmup first. Seems like a strange choice to me. We should perhaps investigate.

amal-ghamdi replied:
Agreed!, feel free to create an issue. At some point I'll look at NUTS checkpointing #404 I can also see if there is something we can do about this. The suggestion I mentioned in that stand-up was to update the original NUTS to have more consistent behavior (and add a flag for legacy behavior). Then we can match NUTSNew with the updated NUTS

nabriis replied:
Added it to #404

amal-ghamdi replied:
great!

nabriis commented on cuqi/experimental/mcmc/_mh.py:
```diff
@@ -33,20 +32,6 @@ def __init__(self, target, proposal=None, scale=1, **kwargs):
     def validate_target(self):
         pass # All targets are valid
 
-    @ProposalBasedSamplerNew.proposal.setter  # TODO. Check if we can refactor this. We can work with a validate_proposal method instead?
```
I don't think we did. But I added the code removed below to validate_proposal in MH and CWMH. And I added required abstract method to implement it.

amal-ghamdi replied:
LGTM

amal-ghamdi commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -67,72 +67,71 @@ class CWMHNew(ProposalBasedSamplerNew):
-        self._scale_temp = self._scale.copy()
+        self._scale = value
+        if isinstance(value, np.ndarray):
+            self._scale_temp = value.copy()
+        else:
+            self._scale_temp = value
```
can we use the property immediately?
```suggestion
            self.scale_temp = value.copy()
        else:
            self.scale_temp = value
```

amal-ghamdi commented on cuqi/experimental/mcmc/_laplace_approximation.py:
```diff
@@ -53,17 +53,14 @@ def __init__(self, target, initial_point=None, maxit=50, tol=1e-4, beta=1e-5, **
 
         super().__init__(target=target, initial_point=initial_point, **kwargs)
 
-        if initial_point is None: #TODO: Replace later with a getter
-            self.initial_point = np.zeros(self.dim)
-            self._samples = [self.initial_point]
-
-        self.current_point = self.initial_point
-        self._acc = [1] # TODO. Check if we need this
-
```
nice clean-up as a result of introducing initialize!

amal-ghamdi commented on cuqi/experimental/mcmc/_pcn.py:
```diff
@@ -64,17 +64,19 @@ def likelihood(self):
-            self._target = value
-            self._loglikelihood = lambda x : self.likelihood.logd(x)
+        if value is not None:
+            if isinstance(value, cuqi.distribution.Posterior):
+                self._target = value
+                self._loglikelihood = lambda x : self.likelihood.logd(x)
+            elif isinstance(value,tuple) and len(value)==2 and \
+                (isinstance(value[0], cuqi.likelihood.Likelihood) or isinstance(value[0], cuqi.likelihood.UserDefinedLikelihood))  and \
+                isinstance(value[1], cuqi.distribution.Distribution):
+                self._target = value
+                self._loglikelihood = lambda x : self.likelihood.logd(x)
+            else:
+                raise ValueError(f"To initialize an object of type {self.__class__}, 'target' need to be of type 'cuqi.distribution.Posterior'.")
         else:
-            raise ValueError(f"To initialize an object of type {self.__class__}, 'target' need to be of type 'cuqi.distribution.Posterior'.")
-        
+            self._target = value
```
target here does not seem call validate_target. Should we instead but the logic of this method in validate_target instead of overwriting the target setter?

nabriis replied:
Ah yes. I guess I was a bit lazy when going over this existing code. I now simplified to 
```python
    def _loglikelihood(self, x):
        return self.likelihood.logd(x)
```
removing the option to define target as tuple ( similar to other simplifications we have done)

nabriis replied:
This also simplified the validate_target method.

amal-ghamdi replied:
nice simplification!

amal-ghamdi commented on cuqi/experimental/mcmc/_rto.py:
```diff
@@ -85,36 +81,38 @@ def data(self):
+                if isinstance(model, np.ndarray) and len(model.shape) == 2:
+                    model = cuqi.model.LinearModel(model)
+
+                # Check model input
+                if not isinstance(model, cuqi.model.LinearModel):
+                    raise TypeError("Model needs to be cuqi.model.LinearModel or matrix")
```
Just a small comment, these few lines (98-100) seem to better fit in `validate_target`

nabriis replied:
I suppose so. although I can't see a nice way to do it. I would actually rather get rid of the tuple option for target similar to the other samplers. I got rid of it now. thoughts?

amal-ghamdi replied:
I think this is a really good idea, to make target only posterior type

amal-ghamdi reviewed (changes requested):
Hi Nicolai, great second round of updates!

It very useful that you introduce the initialize method. I left few comments, feel free to address them as you see fit.

nabriis reviewed (commented):
Hi @amal-ghamdi. Thanks for the comments. I believe i took them all into account. I also added a new test

amal-ghamdi commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -101,39 +111,50 @@ def target(self) -> cuqi.density.Density:
-    def reset(self): # TODO. Issue here. Current point is not reset, and initial point is lost with this reset.
-        self._samples.clear()
-        self._acc.clear()
+    def reset(self):
```
That is a good point (whether parameters is part of checkpointing), maybe it is a good idea to keep track of it somehow.

amal-ghamdi commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -312,13 +337,48 @@ def _call_callback(self, sample, sample_index):
+            if getattr(self, key) is None:
+                raise ValueError(f"Sampler history key {key} is not set after initialization.")
+            
+    def _ensure_initialized(self):
+        """ Ensure the sampler is initialized. If not initialize it. """
+        if not self._is_initialized:
+            self.initialize()
```
yes let's see what @jakobsj suggests

nabriis replied:
We discussed internally. It was agreed to keep wording initialize, but to change name of reset to reinitialize

amal-ghamdi reviewed (approved):
Many thanks @nabriis for this round! Very nice simplification to get rid of tuple in initializing samplers.

I had only one comment for scale temp, feel free to incorporate/or not as you see fit.

amal-ghamdi commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -45,18 +45,32 @@ def __init__(self, target: cuqi.density.Density, initial_point=None, callback=No
         
-        self._samples = [initial_point] # Remove. See #324.
+        self.initial_point = self._default_initial_point
+        self.current_point = self.initial_point
```
👍

jakobsj reviewed (commented):
Many thanks for this extensive PR.  I have gone through it and as far as I can see it achieves the intentions and closes the four listed issues, so that is really great. There is quite a bit of technical details and I have looked through and do not have any specific comments, so my review is here is not a technical review. I see @amal-ghamdi have provided extensive comments that all appear to have been addressed successfully. 

My only comment is in relation to the naming, as requested.  I agree that "initialize" could cause confusion with initial point, especially since the initial point is not part of initialization. The suggestion of "set" is good and would go well with "reset". Also it makes sense to talk about the sampler being in a "set" or "unset" mode, and the "set" method would take it into the "set" mode. I think one thing we have learned is that it is helpful for users (and ourselves) to be quite explicit, and perhaps we can be more explicit than "set", and also avoid a slight confusion since "set" is both the method name and the mode name.

One possibility would be "configure". This makes it more explicit, and like "set" decouples it from the start/initial phase of the sampler. So the method would be "configure" and the mode would be "configured", with a slightly better distinction. "Reset" would then be "reconfigure".

Some things to consider:

1. Do we need the "re" at all?  Could simply "configure" be enough? If calling "configure" on an unconfigured sampler would configure it, and on an already configured sampler would also configure it by replacing the existing configuration by the new one, with the resulting sampler being the same whether or not it was already configured.

2. Whatever the word "configure", "set" or "initialize" I think it is not completely clear that that amounts to setting the state and history. I understand that earlier there was "initialize_state" and "initialize_history", which appear more explicit and thus perhaps clearer? Otherwise the documentation could clarify the definitions and connections between being configured, state and history and distinction vs parameters and other concepts.

nabriis commented:
> Many thanks for this extensive PR. I have gone through it and as far as I can see it achieves the intentions and closes the four listed issues, so that is really great. There is quite a bit of technical details and I have looked through and do not have any specific comments, so my review is here is not a technical review. I see @amal-ghamdi have provided extensive comments that all appear to have been addressed successfully.
> 
> My only comment is in relation to the naming, as requested. I agree that "initialize" could cause confusion with initial point, especially since the initial point is not part of initialization. The suggestion of "set" is good and would go well with "reset". Also it makes sense to talk about the sampler being in a "set" or "unset" mode, and the "set" method would take it into the "set" mode. I think one thing we have learned is that it is helpful for users (and ourselves) to be quite explicit, and perhaps we can be more explicit than "set", and also avoid a slight confusion since "set" is both the method name and the mode name.
> 
> One possibility would be "configure". This makes it more explicit, and like "set" decouples it from the start/initial phase of the sampler. So the method would be "configure" and the mode would be "configured", with a slightly better distinction. "Reset" would then be "reconfigure".
> 
> Some things to consider:
> 
> 1. Do we need the "re" at all?  Could simply "configure" be enough? If calling "configure" on an unconfigured sampler would configure it, and on an already configured sampler would also configure it by replacing the existing configuration by the new one, with the resulting sampler being the same whether or not it was already configured.
> 2. Whatever the word "configure", "set" or "initialize" I think it is not completely clear that that amounts to setting the state and history. I understand that earlier there was "initialize_state" and "initialize_history", which appear more explicit and thus perhaps clearer? Otherwise the documentation could clarify the definitions and connections between being configured, state and history and distinction vs parameters and other concepts.

Hi @jakobsj. As we discussed I have updated the name for reset to reinitialize and kept the other names the same. I also updated the documentation in the classes to make thing more clear (hopefully). Finally I added a nice test for the reinitialization that ensures subclasses implement it consistantly.

jakobsj commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -116,13 +132,12 @@ def target(self, value):
-    def reset(self):
-        """ Reset sampler state and history. """
+    def reinitialize(self):
+        """ Re-initialize the sampler. This resets the state and history and initializes the sampler again. """
```
```suggestion
        """ Re-initialize the sampler. This clears the state and history and initializes the sampler again by setting state and history to their original values. """
```

jakobsj reviewed (changes requested):
I like these changes and I think we are getting close.  I have added a few comments, if you have time it would be great if you can try to address them before we meet today, and then we can progress it further then.

nabriis commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -20,7 +20,16 @@ class SamplerNew(ABC):
 
+    The sampler maintains sets of state and history keys, which are used for features like checkpointing and resuming sampling.
+
+    The state of the sampler represents all variables that are updated (replaced) in a Markov Monte Carlo step, i.e. the current point of the sampler.
```
Good point. Updated.

nabriis commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -61,9 +73,13 @@ def initialize(self):
+
+        # History variables
         self._samples = []
+        self._acc = [ 1 ] # TODO. Check if we need to put 1 here.
```
No. It was moved to the base class

nabriis commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -401,24 +419,32 @@ def __init__(self, target=None, proposal=None, scale=1, **kwargs):
+        self._acc = [ 1 ] # TODO. Check if we need to put 1 here.
+
+        # Parameter variables
+        if self.proposal is None:
```
No. Similar to target and initial point. I move the setting of the default value up to initial_point for clarity

nabriis commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -20,7 +20,16 @@ class SamplerNew(ABC):
+    The state of the sampler represents all variables that are updated (replaced) in a Markov Monte Carlo step, i.e. the current point of the sampler.
+
+    The history of the sampler represents all variables that are updated (appended) in a Markov Monte Carlo step, i.e. the samples and acceptance rates.
+
```
I removed the "parameter" nomenclature and simply refer to any other attribute not part of state and history.

nabriis commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -20,7 +20,16 @@ class SamplerNew(ABC):
+    The history of the sampler represents all variables that are updated (appended) in a Markov Monte Carlo step, i.e. the samples and acceptance rates.
+
+    Subclasses should ensure that any new variables that are updated in a Markov Monte Carlo step are added to the state or history keys.
+
```
Done

nabriis commented:
> I like these changes and I think we are getting close. I have added a few comments, if you have time it would be great if you can try to address them before we meet today, and then we can progress it further then.

I adressed your comments. Thanks. It helped improve the clarity @jakobsj

jakobsj reviewed (approved):
Looks good, thanks!
